### ISSUE TITLE
Is the gradient for Encoder doubled?

### ISSUE THREAD

**Issue by Clement25** [wiseodd/controlled-text-generation#27]
I notice there are two backpropagations for the generator and encoder.

https://github.com/wiseodd/controlled-text-generation/blob/master/train_discriminator.py#L120-L122
https://github.com/wiseodd/controlled-text-generation/blob/master/train_discriminator.py#L130-L132

After the back-propagation of loss G, it runs zero_grad to clear all the grads of the generator in the auto-encoder. However, the encoder is also in the forward path, and its gradient preserves.  Then it computes the encoder loss and back-propagate again.  So the gradient of VAE loss is accumulated twice and the final value is doubled. Is my understanding correct here？

**Comment by Sry2016** [comment]
Hello ! Have you solved this problem?

**Comment by Sry2016** [comment]
You are right.

**Comment by Sry2016** [comment]
You are right. So I think we should add  """trainer_E.zero_grad()“”“ here but without trainer_E.step().
